### PR TITLE
feat: use bound profile provider to call perform.

### PR DESCRIPTION
**Pull request by Jakub-Vacek**
## Description
This PR changes approach to calling SDK. BoundProfileProvider instance is created and used to call perform.
Creation of instance can be divided into two steps:
- resolution of "SF files" we load super.json and local profileAST, mapAST and providerJson (resolveSuperfaceFiles function)
- creation of BoundProfileProvider instance from "SF files"  (createBoundProfileProvider function)

these steps (functions) are used in `prepareSuperface` (naming sucks). This function prepares everything we need to call `perform`. It is easy to mock with `mockBoundProfileProvider` eg. in temporary test file `superface-test-new.test`

This approach can be extended to some (`startRecording` and `endRecording`) functions that are now private in `SuperfaceTest`. That would allow us easier testing (easy mocking + easy unit testing of recording fns)

This approach is, I think:
- easy to test
- easy to read
- unable to use remote files (for now)
- possibly problematic due to OneSdk exports (we need function for preparing BoundProfileProvider exported)

There is a lot of TODOs and thing that can be improved (errors). My main goal was to simplify whole solution and enable reasonable testing.

## Motivation and Context
Using Client that actually calls bind is imo not good approach - it is hard to test and to read. It is also more complicated that it has to be (I believe).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have updated the documentation accordingly. If you are changing **code related to user secrets** you need to really make sure that [security documentation](SECURITY.md) is correct.
- [ ] I have read the **CONTRIBUTION_GUIDE** document.
- [ ] I haven't repeated the code. (DRY)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/superface-test-new.test.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+  writeRecordings: jest.fn(),
+}));
+
+// const DEFAULT_RECORDING_PATH = joinPath(process.cwd(), 'nock');
+
+describe('SuperfaceTest', () => {
+  let superfaceTest: SuperfaceTest;
+
+  afterEach(() => {
+    jest.restoreAllMocks();
```
Comment (freaz) [inline src/superface-test-new.test.ts]: suggestion: Resetting mock in afterEach is here in every (I opened) test, so much be better to turn on mock reseting in global jest configuration. https://jestjs.io/docs/configuration#resetmocks-boolean

**Review by freaz — COMMENTED** [review]
I like where it is heading with all the changes. @Jakub-Vacek kudos for it.

**Comment by martinalbert** [comment]
@Jakub-Vacek @freaz I finished most of what we talked about, so please review 🙏🏻 

I only have to fix not passing tests and maybe figure out how to mock recording better, now that it's abstracted to functions.

File: src/common/format.ts [hunk 2]
```diff
@@ -56,34 +52,37 @@ function isValidPayload(payload: string, match: string): boolean {
...
-    CompleteSuperfaceTestConfig,
-    'profile' | 'provider' | 'useCase'
-  >,
+  profileId: string,
+  providerName: string,
+  usecaseName: string,
+  // sfConfig: Pick<
+  //   CompleteSuperfaceTestConfig,
+  //   'profile' | 'provider' | 'useCase'
+  // >,
```
Comment (Jakub-Vacek) [inline src/common/format.ts]: Remove unused code that I probably left there (sorry) :/

File: src/superface-test.interfaces.ts [hunk 3]
```diff
@@ -9,8 +8,6 @@ import {
 } from '@superfaceai/one-sdk';
 import { Definition } from 'nock/types';
 
-import { ISuperfaceClient } from './superface/client';
-
 export interface SuperfaceTestConfigPayload {
```
Comment (Jakub-Vacek) [inline src/superface-test.interfaces.ts]: I would named this SuperfaceTestConfig. Payload feels unnecessary.

File: src/superface-test.test.ts [hunk 4]
```diff
@@ -1,47 +1,31 @@
 import { ApiKeyPlacement, SecurityType } from '@superfaceai/ast';
```
Comment (Jakub-Vacek) [inline src/superface-test.test.ts:1]: Tests looks much better now 👍

File: src/superface-test.test.ts [hunk 5]
```diff
@@ -510,6 +570,27 @@ describe.skip('SuperfaceTest', () => {
         mocked(exists).mockResolvedValueOnce(true);
         mocked(readFileQuiet).mockResolvedValueOnce('[]');
         mocked(matchWildCard).mockReturnValueOnce(false);
+        mocked(prepareSuperface).mockResolvedValue({
```
Comment (Jakub-Vacek) [inline src/superface-test.test.ts]: mocking of prepareSuperface can be probably move to some top level function to get rid of repeating it in every case.

File: src/superface/config/create-bound-profile-provider.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  NodeCrypto,
+  NodeFetch,
+  NodeFileSystem,
+  NodeLogger,
+  NodeTimers,
+  profileAstId,
+  resolveSecurityConfiguration,
+  ServiceSelector,
+} from '@superfaceai/one-sdk';
+// TODO: export from SDK?
```
Comment (Jakub-Vacek) [inline src/superface/config/create-bound-profile-provider.ts]: You can remove the TODO :)

File: src/superface/config/prepare-ast.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  } else {
+    const profileSettings = superJson.profiles[profile];
+
+    if (!profileSettings || !('file' in profileSettings)) {
+      throw new ProfileUndefinedError(profile);
+    }
+
+    // TODO: resolve path
+    const profilePath = profileSettings.file;
+    const content = await fileSystem.readFile(profilePath);
```
Comment (Jakub-Vacek) [inline src/superface/config/prepare-ast.ts]: This TODO means that something like this:

resolvePath(dirname(superJsonPath), path)

should be used to resolve the file. Try it in station just to be sure. I am not saying that you need to change just make sure it works :)

File: src/superface/config/prepare-ast.ts [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+    if (content.isErr()) {
+      throw content.error;
+    }
+
+    // TODO: make parser optional
```
Comment (Jakub-Vacek) [inline src/superface/config/prepare-ast.ts]: Remove TODO

File: src/superface/config/prepare-ast.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    if (profilePath.endsWith('.supr')) {
+      return parseProfile(new Source(content.value, profilePath));
+    } else {
+      return assertProfileDocumentNode(JSON.parse(content.value));
+    }
+  }
+}
+
+export async function getMapAst(
```
Comment (Jakub-Vacek) [inline src/superface/config/prepare-ast.ts:46]: Same stuff as with profile

File: src/superface/config/prepare-provider-json.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+): Promise<ProviderJson> {
+  const providerName =
+    provider instanceof Provider ? provider.configuration.name : provider;
+  const providerSettings = superJson.providers[providerName];
+
+  if (!providerSettings || !providerSettings.file) {
+    throw new ProviderJsonUndefinedError(providerName);
+  }
+
+  // superJson.resolvePath(providerSettings.file);
```
Comment (Jakub-Vacek) [inline src/superface/config/prepare-provider-json.ts]: Same as with map :)

**Review by Jakub-Vacek — COMMENTED** [review]
Added some comments on the small stuff. Overall looks good.

File: src/superface/config/prepare-ast.ts (hunk 8 above)
Comment (martinalbert) [reply src/superface/config/prepare-ast.ts]: I added this, I can add dynamic import and call the fallback without parser if import fails. The question is, what will be fallback. 😄

Comment (Jakub-Vacek) [reply src/superface/config/prepare-ast.ts]: Fall should be error saying that you need to compile and there also should be extension check. But I am ok with keeping parser in deps.

File: src/superface/config/prepare-ast.ts (hunk 7 above)
Comment (martinalbert) [reply src/superface/config/prepare-ast.ts]: After testing in station, I figured out that `dirname` is not necessary as `detectSuperJson` returns path to directory, not the exact path, so I used `resolvePath(superJsonPath, path)`

File: src/superface/mock/timers.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import { ITimeout, ITimers } from '@superfaceai/one-sdk';
+
+export class MockTimeout implements ITimeout {
+  constructor(public value: number) {}
+}
+
+export class MockTimers implements ITimers {
```
Comment (Jakub-Vacek) [inline src/superface/mock/timers.ts:7]: Why didn't you just reused mocks from SDK?

Comment (martinalbert) [reply src/superface/mock/timers.ts:7]: Because it's not exported from sdk as far as I know 😄

**Comment by Jakub-Vacek** [comment]
It looks ok (I checked files that were changed after my last review). Please resolve conflicts before final review.

**Pull request merged**